npm/private/npm_translate_lock_state.bzl

```diff
@@ -208,19 +209,51 @@ def _init_root_package(priv, rctx, attr, label_store):
             fail("root_package cannot be overridden if there are pnpm workspace packages specified")
         priv["root_package"] = attr.root_package
 
-    # Load a declared root package.json
+def _init_workspace(priv, rctx, attr, label_store, is_windows):
+    root_package_json = {}
+
+    # Load pnpm settings from root package.json for pnpm <= v9.
     if label_store.has("package_json_root"):
+        # Load a declared root package.json
         root_package_json_path = label_store.path("package_json_root")
-        priv["root_package_json"] = json.decode(rctx.read(root_package_json_path))
+        root_package_json = json.decode(rctx.read(root_package_json_path))
     elif "." in priv["importers"].keys():
         # Load an undeclared package.json derived from the root importer in the lockfile
         label_store.add_sibling("lock", "package_json_root", PACKAGE_JSON_FILENAME)
         root_package_json_path = label_store.path("package_json_root")
-        priv["root_package_json"] = json.decode(rctx.read(root_package_json_path))
-    else:
-        # if there is no root importer that means there is no root package.json to read; pnpm allows
-        # you to just have a pnpm-workspaces.yaml at the root and no package.json at that location
-        priv["root_package_json"] = {}
+        root_package_json = json.decode(rctx.read(root_package_json_path))
+
+    priv["pnpm_settings"] = root_package_json.get("pnpm", {})
+
+    # Read settings from pnpm-workspace.yaml for pnpm v10+ (NOTE: pnpm 9-10+ has lockfile version 9).
+    # Support scenario where pnpm-lock.yaml was never parsed and "lock_version" is not set.
+    if priv.get("lock_version", 0) >= 9.0 and label_store.has("pnpm_workspace"):
+        pnpm_workspace_path = label_store.path("pnpm_workspace")
+        if is_windows or utils.exists(rctx, pnpm_workspace_path):
```

Comment:
i feel like all these `is_windows` may not really be correct but we'll never know :) Hopefully we can come back to fixing these using `path.exists` properly

Reply:
Yeah I'm curious if we can clean this up in 3.x?